torch/nn/parallel/distributed.py

```diff
@@ -599,3 +653,20 @@ def _passing_sync_batchnorm_handle(self, module_copies):
                     assert self.device_type != 'cpu', "SyncBatchNorm layers only work with GPU modules"
                     layer._specify_ddp_gpu_num(
                         len(self.device_ids) if self.device_ids else 1)
+
+    def _check_comm_hook(self, hook):
+        assert callable(hook), "Communication hook must be callable."
+
+        sig = inspect.signature(hook)
+        assert (sig.parameters['state'].annotation == inspect._empty or
+                sig.parameters['state'].annotation == object), (
+            "Communication hook: state annotation is not object."
+        )
+        assert (sig.parameters['bucket'].annotation == inspect._empty or
+                sig.parameters['bucket'].annotation == dist.GradBucket), (
+            "Communication hook: bucket annotation is not dist.GradBucket."
+        )
+        assert (sig.return_annotation == inspect._empty or
+                sig.return_annotation == torch.futures.Future), (
+            "Communication hook: return annotation is not torch.futures.Future."
+        )
```

Comment:
I think if there are no annotations, then we won't have any typechecking? I think that's probably fine, because typical patterns I've seen with Python is to not enforce typechecking rigorously and just let the exceptions handle it when they happen.

Reply:
Without any annotations, no type check will be done.
Adding an exception message following https://github.com/pytorch/pytorch/pull/40848#discussion_r452343549  will be a better internal check. Still we are not enforcing users to use annotations, so I think it is fine to keep annotation checks just for hook and return.